contracts/root/predicates/MarketplacePredicate.sol

```diff
@@ -0,0 +1,282 @@
+pragma solidity ^0.5.2;
+
+import { BytesLib } from "../../common/lib/BytesLib.sol";
+import { Common } from "../../common/lib/Common.sol";
+import { ECVerify } from "../../common/lib/ECVerify.sol";
+import { Math } from "openzeppelin-solidity/contracts/math/Math.sol";
+import { RLPEncode } from "../../common/lib/RLPEncode.sol";
+import { RLPReader } from "solidity-rlp/contracts/RLPReader.sol";
+
+import { IPredicate, PredicateUtils } from "./IPredicate.sol";
+import { Registry } from "../../common/Registry.sol";
+import { IWithdrawManager } from "../withdrawManager/IWithdrawManager.sol";
+
+contract MarketplacePredicate is PredicateUtils {
+  using RLPReader for bytes;
+  using RLPReader for RLPReader.RLPItem;
+
+  // 0xe660b9e4 = keccak256('executeOrder(bytes,bytes,bytes32,uint256,address)').slice(0, 4)
+  bytes4 constant EXECUTE_ORDER_FUNC_SIG = 0xe660b9e4;
+
+  IWithdrawManager internal withdrawManager;
+  Registry internal registry;
+
+  struct ExecuteOrderData {
+    bytes data1;
+    bytes data2;
+    bytes32 orderId;
+    uint256 expiration;
+    address taker;
+  }
+
+  struct Order {
+    address token;
+    bytes sig;
+    uint256 amount;
+  }
+
+  struct ExitTxData {
+    // token1 and amount1 should correspond to what the exitor (msg.sender) signed over
+    uint256 amount1;
+    uint256 amount2;
+    address token1;
+    address token2;
+    address counterParty;
+    bytes32 txHash;
+  }
+
+  struct ReferenceTxData {
+    uint256 closingBalance;
+    uint256 age;
+    address childToken;
+    address rootToken;
+  }
+
+  constructor(address _withdrawManager, address _registry)
+    public
+  {
+    withdrawManager = IWithdrawManager(_withdrawManager);
+    registry = Registry(_registry);
+  }
+
+  function startExit(bytes calldata data, bytes calldata exitTx)
+    external
+  {
+    ExitTxData memory exitTxData = processExitTx(exitTx, withdrawManager.networkId());
+    RLPReader.RLPItem[] memory referenceTx = data.toRlpItem().toList();
+    (address predicate, bytes memory preState) = abi.decode(referenceTx[0].toBytes(), (address, bytes));
+    require(
+      uint8(registry.predicates(predicate)) != 0,
```

Comment:
why are we just type casting the whole struct? it would be better if we use `registry.predicates(predicate)._type` and then typecast.